### PR TITLE
Porting to ros2

### PR DESCRIPTION
**Pull request by RoboticsYY**
This PR intends to port this package to ROS2.

#### Run test on Ubuntu 20.04 with ROS2 Foxy
- Install dependencies: `sudo apt install mongodb libmongoclient-dev`
- `mkdir -p ws_ros2/src && cd ws_ros2/src`
- `git clone -b pr-ros2 https://github.com/RoboticsYY/warehouse_ros`
- `git clone -b pr-porting_to_ros2 https://github.com/RoboticsYY/warehouse_ros_mongo `
- `cd .. && source /opt/ros/foxy/setup.bash`
- `colcon build --cmake-args -DCMAKE_BUILD_TYPE=Release && source install/setup.bash`
- `colcon test --packages-select warehouse_ros_mongo`

### CODE REVIEW

File: warehouse_ros_mongo/mongo_wrapper_ros.py [hunk 1]
```diff
@@ -30,55 +32,61 @@ def print_help_message():
...
         os.remove(lock_file)
     check_call(['mongodb', 'mongod', '--repair', '--dbpath', dbpath])
-    rospy.loginfo("  Successfully repaired.")
+    node.get_logger().info("  Successfully repaired.")
     sys.exit(0)
 
 # The defaults here should match the ones used by each client library
-port = rospy.get_param('warehouse_port', 27017)
-host = rospy.get_param('warehouse_host', 'localhost')
+port = node.get_parameter_or('warehouse_port', Parameter('', Parameter.Type.INTEGER, 27017))
```
Comment (henningkayser) [inline warehouse_ros_mongo/mongo_wrapper_ros.py]: Shouldn't this use the new port?

File: warehouse_ros_mongo/mongo_wrapper_ros.py [hunk 2]
```diff
@@ -30,55 +32,61 @@ def print_help_message():
...
+        node.get_logger().info("Ignoring mongod's non-standard return code of 12")
     else:
-        rospy.logerr("Mongo process exited with error code {0}".format(e.returncode))
+        node.get_logger().error("Mongo process exited with error code {0}".format(e.returncode))
 except OSError as e:
-    rospy.logerr("Execution failed: %s", e)
+    node.get_logger().error("Execution failed: {}".format(e))
+
+rclpy.spin(node)
+node.destroy_node()
```
Comment (henningkayser) [inline warehouse_ros_mongo/mongo_wrapper_ros.py]: Just curious, do we really have to call these explicitly?

File: test/test_mongo_helpers.h [hunk 3]
```diff
@@ -64,25 +64,25 @@ inline warehouse_ros_mongo::MongoQuery& downcastQuery(warehouse_ros::Query::Cons
 template <class T>
 ostream& operator<<(ostream& str, const warehouse_ros::MessageWithMetadata<T>& s)
 {
-  const T& msg = s;
-  str << "Message: " << msg;
+  // const T& msg = s;
```
Comment (henningkayser) [inline test/test_mongo_helpers.h:68]: There should really be a way to pretty-print messages. Please add a TODO to fix this

**Comment by henningkayser** [comment]
clang-format needs to be fixed

File: warehouse_ros_mongo/mongo_wrapper_ros.py (hunk 1 above)
Comment (RoboticsYY) [reply warehouse_ros_mongo/mongo_wrapper_ros.py]: Changed this to the new port number that aligns with the launch file.

File: warehouse_ros_mongo/mongo_wrapper_ros.py (hunk 2 above)
Comment (RoboticsYY) [reply warehouse_ros_mongo/mongo_wrapper_ros.py]: No, this is not necessary. These lines are removed.

File: test/test_mongo_helpers.h (hunk 3 above)
Comment (RoboticsYY) [reply test/test_mongo_helpers.h:68]: A TODO comment is added here.

File: src/query_results.cpp [hunk 4]
```diff
@@ -83,15 +85,15 @@ std::string MongoResultIterator::message() const
...
-  ROS_ASSERT(f.exists());
+  rcpputils::assert_true(f.exists());
   std::stringstream ss(std::ios_base::out);
   f.write(ss);
   return ss.str();
 }
 
 mongo::BSONObj MongoResultIterator::metadataRaw() const
 {
-  ROS_ASSERT(next_);
```
Comment (henningkayser) [inline src/query_results.cpp:94]: Wouldn't this be a use case of `RCLCPP_FATAL_EXPRESSION`?

**Review by henningkayser — APPROVED** [review]
Looks good to me

File: src/query_results.cpp (hunk 4 above)
Comment (RoboticsYY) [reply src/query_results.cpp:94]: Thanks for pointing this out! Yes, I think `RCLCPP_FATAL_EXPRESSION` is a better option here, which can provide more debug info. Changed to `RCLCPP_FATAL_EXPRESSION`.

**Pull request merged**